Framework/API/src/Algorithm.cpp

```diff
@@ -1616,6 +1620,153 @@ void Algorithm::setAlgStartupLogging(const bool enabled) {
+
+  const auto storageModes = getInputWorkspaceStorageModes();
+  const auto executionMode = getParallelExecutionMode(storageModes);
+  if (executionMode == Parallel::ExecutionMode::Invalid) {
+    std::string error("Algorithm does not support execution with input "
+                      "workspaces of the following storage types: " +
+                      Parallel::toString(storageModes) + ".");
+    getLogger().error() << error << "\n";
+    throw(std::runtime_error(error));
+  }
+  if (executionMode == Parallel::ExecutionMode::Serial) {
+    std::string error(Parallel::toString(executionMode) +
+                      " is not a valid *parallel* execution mode.");
+    getLogger().error() << error << "\n";
+    throw(std::runtime_error(error));
+  }
+  return executionMode;
+}
+
+/** Get map of storage modes of all input workspaces.
+ *
+ * The key to the name is the property name of the respective workspace. */
+std::map<std::string, Parallel::StorageMode>
+Algorithm::getInputWorkspaceStorageModes() const {
+  std::map<std::string, Parallel::StorageMode> map;
+  for (const auto &wsProp : m_inputWorkspaceProps) {
+    // This is the reverse cast of what is done in cacheWorkspaceProperties(),
+    // so it should never fail.
+    const Property &prop = dynamic_cast<Property &>(*wsProp);
+    // Check if we actually have that input workspace
+    if (wsProp->getWorkspace())
+      map.emplace(prop.name(), wsProp->getWorkspace()->storageMode());
+  }
+  return map;
+}
+
+/** Get correct execution mode based on input storage modes for an MPI run.
+ *
+ * The default implementation returns ExecutionMode::Invalid. Classes inheriting
+ * from Algorithm can re-implement this if they support execution with multiple
+ * MPI ranks. May not return ExecutionMode::Serial, because that is not a
+ * "parallel" execution mode. */
+Parallel::ExecutionMode Algorithm::getParallelExecutionMode(
+    const std::map<std::string, Parallel::StorageMode> &storageModes) const {
+  UNUSED_ARG(storageModes)
+  // By default no parallel execution is possible.
+  return Parallel::ExecutionMode::Invalid;
+}
+
+/// Sets up skipping workspace validation on non-master ranks for
+/// StorageMode::MasterOnly.
+void Algorithm::setupSkipValidationMasterOnly() {
+  // If workspaces have StorageMode::MasterOnly, validation on non-master ranks
+  // would usually fail. Therefore, WorkspaceProperty needs to skip validation.
+  // Thus, we must notify it whether or not it is on the master rank or not.
+  if (communicator().rank() != 0)
+    for (auto *prop : getProperties())
+      if (auto *wsProp = dynamic_cast<IWorkspaceProperty *>(prop))
+        wsProp->setIsMasterRank(false);
+}
+
+/// Returns a const reference to the (MPI) communicator of the algorithm.
+const Parallel::Communicator &Algorithm::communicator() const {
+  return *m_communicator;
+}
+
+/// Sets the (MPI) communicator of the algorithm.
+void Algorithm::setCommunicator(const Parallel::Communicator &communicator) {
```

Comment:
This is currently never used. We currently just use the default constructed `Communicator` on the `Algorithm`?